File: STRM/src/main/java/hcs/offloading/strm/Dispatcher.java
package hcs.offloading.strm;

import static hcs.offloading.strm.PatchMixer.CONTINUE_PACKING;

import android.graphics.Bitmap;
import android.graphics.Rect;
import android.util.Pair;

import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import hcs.offloading.strm.config.DispatcherConfig;
import hcs.offloading.strm.config.RoIExtractorConfig;
import hcs.offloading.strm.datatypes.BoundingBox;
import hcs.offloading.strm.datatypes.Frame;

public class Dispatcher extends Consumer<Frame> {
    private static final String TAG = Dispatcher.class.getName();

    private int mCountMixedFrameInference = Integer.MAX_VALUE;
    private Frame mPrevFrame;
    private List<BoundingBox> mPrevResults;

    private final DispatcherConfig mConfig;
    private final Map<Integer, Frame> mFrames = new HashMap<>();

    private final RoIExtractor mRoIExtractor;
    private final RoIPrioritizer mRoIPrioritizer;
    private final ResizeProfile mResizeProfile;
    private final InferenceEngine mInferenceEngine;
    private final PatchMixer mPatchMixer;

    Dispatcher(DispatcherConfig config,
               RoIExtractorConfig roIExtractorConfig,
               ResizeProfile resizeProfile,
               RoIPrioritizer roIPrioritizer,
               InferenceEngine inferenceEngine,
               PatchMixer patchMixer) {
        super(TAG, config.MAX_QUEUE_SIZE, null);
        mConfig = config;
        mRoIExtractor = new RoIExtractor(roIExtractorConfig);
        mRoIPrioritizer = roIPrioritizer;
        mResizeProfile = resizeProfile;
        mInferenceEngine = inferenceEngine;
        mPatchMixer = patchMixer;
    }

    @Override
    public void process(Frame currFrame) throws InterruptedException {
        assert mPrevFrame == null || mPrevFrame.frameIndex + 1 == currFrame.frameIndex;
        /* Cases
         * 1. Full frame inference
         * 2. Mixed frame inference
         *   2.1. If only single frame packed into mixed frame, use inference result of the packed frame.
         *   2.2. Else if multiple frames packed into mixed frame, exclude the last packed frame
         *        and re-pack the frame into next mixed frame.
         */
        synchronized (mFrames) {
            mFrames.put(currFrame.frameIndex, currFrame);
            mFrames.notifyAll();
        }
        if (mCountMixedFrameInference >= mConfig.FULL_INFERENCE_INTERVAL) {
            mCountMixedFrameInference = 0;
            Bitmap copiedBitmap = currFrame.bitmap.copy(currFrame.bitmap.getConfig(), false);
            int handle = mInferenceEngine.enqueue(copiedBitmap, true);
            List<BoundingBox> results = mInferenceEngine.getResults(handle);
            currFrame.setResults(results);
            setPrevBoxesForOpticalFlowRoI(results, false);
        } else {
            List<BoundingBox> prevResults = getPrevResults();
            mRoIExtractor.process(new Pair<>(new Pair<>(mPrevFrame, currFrame), prevResults));
            currFrame.sortRoIs(Comparator.comparingInt(mRoIPrioritizer::priority));
            currFrame.getRoIs().forEach(roi -> roi.resize(
                    mResizeProfile.getScale(roi.labelName, roi.location.width(), roi.location.height(), roi.minOriginLength)));
            int lastPackedIndex = mPatchMixer.tryPackAndEnqueueMixedFrame(currFrame);
            if (lastPackedIndex == CONTINUE_PACKING) {
                setPrevBoxesForOpticalFlowRoI(currFrame.getOpticalFlowRoIs().stream()
                        .map(roi -> new BoundingBox(roi.location, 1f, roi.labelName))
                        .collect(Collectors.toList()), true);
            } else {
                mCountMixedFrameInference++;
                boolean isCurrFrameExcludedFromMixedFrame = lastPackedIndex == currFrame.frameIndex - 1;
                if (isCurrFrameExcludedFromMixedFrame) {
                    process(currFrame);
                }
            }
        }
        mPrevFrame = currFrame;
    }

    void setPrevBoxesForOpticalFlowRoI(List<BoundingBox> prevBoxes, boolean isRoI) {
        synchronized (this) {
            if (isRoI) { // Optical Flow RoIs
                mPrevResults = prevBoxes;
            } else { // Inference Results
                mPrevResults = prevBoxes.stream()
                        .map(box -> new BoundingBox(new Rect(
                                box.location.left - mConfig.ROI_PADDING,
                                box.location.top - mConfig.ROI_PADDING,
                                box.location.right + mConfig.ROI_PADDING,
                                box.location.bottom + mConfig.ROI_PADDING),
                                box.confidence, box.labelName))
                        .collect(Collectors.toList());
            }
            notifyAll();
        }
    }

    private List<BoundingBox> getPrevResults() throws InterruptedException {
        if (!mRoIExtractor.useOpticalFlowRoIs()) {
            return null;
        }
        synchronized (this) {
            while (mPrevResults == null) {
                wait();
            }
            List<BoundingBox> prevResults = mPrevResults;
            mPrevResults = null;
            return prevResults;
        }
    }

    public Frame getResults(int frameIndex) throws InterruptedException {
        Frame frame;
        synchronized (mFrames) {
            frame = mFrames.remove(frameIndex);
            while (!isClosed() && frame == null) {
                mFrames.wait();
                frame = mFrames.remove(frameIndex);
            }
        }
        if (frame == null) {
            return null;
        } else {
            frame.waitForResults();
            return frame;
        }
    }

    @Override
    public void close() {
        super.close();
        synchronized (mFrames) {
            mFrames.notifyAll();
        }
    }
}

File: STRMCPP/src/main/cpp/src/Dispatcher.cpp
#include "strm/Dispatcher.hpp"

#include <cassert>

#include "strm/Log.hpp"

namespace rm {

void Dispatcher::process(Frame*& currFrame) {
  LOGD("Dispatcher::process %d %d",
       currFrame != nullptr ? currFrame->frameIndex : -1,
       mPrevFrame != nullptr ? mPrevFrame->frameIndex : -1);
  assert(currFrame != nullptr && mPrevFrame == nullptr || mPrevFrame->frameIndex + 1 == currFrame->frameIndex);
  /* Cases
   * 1. Full frame inference
   * 2. Mixed frame inference
   *   2.1. If multiple frames from the source stream packed into mixed frame,
   *        PatchMixer will exclude the last packed frame from the packed frames.
   *        We have to re-pack the frame into next mixed frame.
   *   2.2. Else wait for mixed frame result and continue to process next frame.
   */
  mFrames.insert(std::make_pair(currFrame->frameIndex, std::unique_ptr<Frame>(currFrame)));
  if (mCountMixedFrameInference >= mConfig.FULL_INFERENCE_INTERVAL) {
    mCountMixedFrameInference = 0;
    mUseInferenceResults = true;
    int handle = mInferenceEngine->enqueue(currFrame->mat->clone(), true);
    std::vector<BoundingBox> results = mInferenceEngine->getResults(handle);
    currFrame->boxes.insert(currFrame->boxes.end(), results.begin(), results.end());
    notifyResults();
  } else {
    std::vector<BoundingBox> prevResults = getPrevBoxes();
    mRoIExtractor.process(std::make_pair(std::make_pair(mPrevFrame, currFrame), prevResults));
    std::sort(currFrame->rois.begin(), currFrame->rois.end(), [this](const RoI& lhs, const RoI& rhs) -> bool {
      return mRoIPrioritizer->priority(lhs) < mRoIPrioritizer->priority(rhs);
    });
    for (auto& roi : currFrame->rois) {
      roi.scale = mResizeProfile->getScale(roi.labelName, roi.location.width(),
                                           roi.location.height(), roi.minOriginLength);
    }
    PatchMixer::Status status = mPatchMixer->tryPackAndEnqueueMixedFrame(currFrame);
    if (status == PatchMixer::CONTINUE_PACKING) {
      mUseInferenceResults = false;
    } else if (status == PatchMixer::FINISHED) {
      mCountMixedFrameInference++;
      mUseInferenceResults = true;
    } else if (status == PatchMixer::FINISHED_AND_PROCESS_LAST_FRAME_AGAIN) {
      mCountMixedFrameInference++;
      mUseInferenceResults = true;
      process(currFrame);
    } else {
      // TODO: Error handling
    }
  }
  mPrevFrame = currFrame;
}

std::vector<BoundingBox> Dispatcher::getPrevBoxes() {
  LOGD("Dispatcher::getPrevBoxes");
  std::vector<BoundingBox> prevResults;
  if (!mRoIExtractor.useOpticalFlowRoIs()) {
    return prevResults;
  }
  std::unique_lock<std::mutex> lock(mtx);
  if (mUseInferenceResults) {
    cv.wait(lock, [this](){
      return mPrevFrame->isResultReady.load();
    });
    for (const BoundingBox& box : mPrevFrame->boxes) {
      prevResults.emplace_back(Rect(
              box.location.left - mConfig.ROI_PADDING,
              box.location.top - mConfig.ROI_PADDING,
              box.location.right + mConfig.ROI_PADDING,
              box.location.bottom + mConfig.ROI_PADDING), box.confidence, box.labelName);
    }
  } else {
    for (const RoI& roi : mPrevFrame->opticalFlowRoIs) {
      prevResults.emplace_back(roi.location, 1.0, roi.labelName);
    }
  }
  return prevResults;
}

void Dispatcher::notifyResults() {
  LOGD("Dispatcher::notifyResults");
  cv.notify_all();
}

std::vector<BoundingBox> Dispatcher::getResults(int frameIndex) {
  LOGD("Dispatcher::getResults Start %d", frameIndex);
  std::unique_lock<std::mutex> lock(mtx);
  cv.wait(lock, [this, &frameIndex]() {
    return mFrames.find(frameIndex) != mFrames.end() && mFrames.at(frameIndex)->isResultReady.load();
  });
  LOGD("Dispatcher::getResults End   %d", frameIndex);
  return mFrames.at(frameIndex)->boxes;
}

void Dispatcher::enqueue(Frame* item) {
  LOGD("Dispatcher::enqueue start %d", item->frameIndex);
  std::unique_lock<std::mutex> lock(mItemsMtx);
  mItemsCV.wait(lock, [this] {
    return mItems.size() < mMaxNumItems;
  });
  mItems.push(item);
  lock.unlock();
  mItemsCV.notify_all();
  LOGD("Dispatcher::enqueue end   %d", item->frameIndex);
}

Frame* Dispatcher::takeItem() {
  LOGD("Dispatcher::takeItem start");
  std::unique_lock<std::mutex> lock(mItemsMtx);
  mItemsCV.wait(lock, [this] {
    return !mItems.empty();
  });
  Frame* item = mItems.front();
  mItems.pop();
  lock.unlock();
  mItemsCV.notify_all();
  LOGD("Dispatcher::takeItem end %d", item->frameIndex);
  return item;
}

} // namespace rm
